introduce getPeerCertificates, fixes #13299

royneary opened the pull request:
getPeerCertificates retrieves the verified certificate chain of the peer
we are connected to through an SSL-wrapped Socket/AsyncSocket. This
introduces the new type ``Certificate`` which stores a DER-encoded X509 certificate.

Araq commented:
Add a ``.since`` annotation, a changelog entry and rebase since Azure pipelines got the flu and didn't test anything.

royneary committed:
make i2d_X509 and d2i_X509 always available

i2d_X509 and d2i_X509 have been available in all versions of OpenSSL, so
make them available even if nimDisableCertificateValidation is set.

royneary commented:
Done.

In the meanwhile 5b854442448af96d57135ba7328b0c21f1f80f40 introduced the wrapper ``proc i2d_X509*(cert: PX509): string`` which is useful for ``getPeerCertificates``. Unfortunately it was defined in a ``when not defined(nimDisableCertificateValidation)`` block. My code is not about certificate validation, so it should compile independently from the ``nimDisableCertificateValidation`` flag.

To solve this, I moved ``i2d_X509`` out of the block. This should be ok since the reason for ``nimDisableCertificateValidation`` (that some symbols are not available in older OpenSSL versions) does not apply to ``i2d_X509`` which has been "available in all versions of SSLeay and OpenSSL" [1].

For consistency I moved the reverse proc ``d2i_X509`` too.

[1] https://linux.die.net/man/3/i2d_x509

royneary committed:
introduce getPeerCertificates, fixes #13299

getPeerCertificates retrieves the verified certificate chain of the peer
we are connected to through an SSL-wrapped Socket/AsyncSocket. This
introduces the new type Certificate which stores a DER-encoded X509 certificate.
